strawberryfields/engine.py

```diff
@@ -540,17 +548,19 @@ def run(self, program: Program, *, compile_options=None, **kwargs) -> Optional[R
             ``None`` otherwise
         """
         job = self.run_async(program, compile_options=compile_options, **kwargs)
+        log.info("Submitted job {}".format(job.id))
         try:
             while True:
                 job.refresh()
                 if job.status == "complete":
                     return job.result
                 if job.status == "failed":
                     log.warning(
-                        "The remote job failed due to an internal server error; "
-                        "please try again."
+                        "The remote job {} failed due to an internal server error; "
+                        "please try again.".format(job.id)
                     )
                     return None
+
                 time.sleep(self.POLLING_INTERVAL_SECONDS)
         except KeyboardInterrupt:
```

Comment:
I wonder whether the KeyboardInterrupt should not be re-raised after having cancelled the job. Usually a user script has some other commands after running the job, that then fail because the returned object is None.

Reply:
sorry for this remark, I guess it is off-topic.

Reply:
That sounds reasonable to me, though I guess it would more so serve explanatory purposes. This would go from 
```python
AttributeError                            Traceback (most recent call last)
<ipython-input-1-400e2637265b> in <module>
     16     ops.MeasureFock() | q
     17 result = eng.run(prog, shots=1000)
---> 18 result.cutoff_dim

AttributeError: 'NoneType' object has no attribute 'cutoff_dim'
```
to having
```python
    559         except KeyboardInterrupt:
    560             self._connection.cancel_job(job.id)
--> 561             raise KeyboardInterrupt()
    562 
    563     def run_async(self, program: Program, *, compile_options=None, **kwargs) -> Job:

KeyboardInterrupt: 
```
Or something comparable, based on any message raised there.

Would it then also make sense to raise an error when the job failed?https://github.com/XanaduAI/strawberryfields/blob/f6ac22b00c8a085e0fbd75d3a19b631e634e4c0f/strawberryfields/engine.py#L548


 Right now there is a warning. However, following these lines of thoughts, if `run` doesn't return a `Result` but `None`, then an error will be raised anyways due to operations in the script that were meant to be carried out on the `Result` instance.

Reply:
I agree, things should be consistent, so lets drop this here, as there would be many more place to touch if one wanted to fix this. I guess the choice to return `None` rather than raising an exception was made at some point for good reasons.

Reply:
Just throwing out ideas here... I think it might make sense to raise a `JobCancelled` exception or something similar, or just exit outright (by re-raising as suggested, or `sys.exit`, etc.). The use case I have in mind for the former is if the user, for some reason, wants to continue execution after cancelling a single job. If they choose not (or don't know) to catch the `JobCancelled`, then the script terminates anyway; if they catch it, then they can presumably write their own code to handle the `None`. And I guess the same idea would apply to failed jobs.

Reply:
From the user perspective, testing for `result is None` or catching the exception provides roughly the same functionality I guess. Except catching an exception would allow more fine-grained flow control than checking for None, e. g. different reactions to `JobCancelled` and `JobFailed` errors or the like. So I would favor an exception over None, in the spirit of 

> Errors should never pass silently.